hxmath/math/DualQuaternion.hx

```diff
@@ -0,0 +1,622 @@
+    }
+    
+    /**
+     * divide by constant
+     * 
+     * @param a
+     * @param v
+     * @return       a/v
+     */
+    @:op(A / B)
+    public static inline
+    function divide(a:DualQuaternion, v:Float):DualQuaternion
+    {
+        return a * ( 1 / v );
+    }
+    
+    /**
+     * applyScale
+     * 
+     * @param a
+     * @return       scaled 
+     */
+    public inline function applyScale(s:Float):DualQuaternion
+    {
+        var cloned = clone();
+        cloned = cloned*DualQuaternion.scalar( s );
+        return cloned;
+    }
+    
+    /**
+     * Add two dual quaternions.
+     * 
+     * @param a
+     * @param b
+     * @return      a + b
+     */
+    @:op(A + B)
+    public static inline function add(a:DualQuaternion, b:DualQuaternion):DualQuaternion
+    {
+        return a.clone()
+            .addWith(b);
+    }
+    
+    /**
+     * Add a quaternion in place.
+     * 
+     * @param a
+     * @return      self_i += a_i
+     */
+    public inline function addWith(a:DualQuaternion):DualQuaternion
+    {
+        return new DualQuaternion( this.real + a.real, this.dual + a.dual );
+    }
+    
+    /**
+     * Subtract one dual quaternion from another.
+     * 
+     * @param a
+     * @param b
+     * @return      a - b
+     */
+    @:op(A - B)
+    public static inline function subtract(a:DualQuaternion, b:DualQuaternion):DualQuaternion
+    {
+        return a.clone()
+            .subtractWith(b);
+    }
+    
+    /**
+     * Subtract a dual quaternion in place.
+     *
+     * @param a
+     * @return      self_i -= a_i
+     */
+    public inline function subtractWith(a:DualQuaternion):DualQuaternion
+    {
+        return new DualQuaternion( this.real - a.real, this.dual - a.dual );
+    }
+    
+    /**
+     * Multiply two dual quaternions.
+     * 
+     * @param a
+     * @param b
+     * @return      a * b
+     */
+    @:op(A * B)
+    public static inline
+    function multiply( q1: DualQuaternion, q2: DualQuaternion ):DualQuaternion
+    {
+        return new DualQuaternion(q1.real * q2.real, q1.real * q2.dual + q1.dual * q2.real);
+    }
+    
+    private static inline function get_zero(): DualQuaternion
+    {
+        return new DualQuaternion(Quaternion.zero, Quaternion.zero);
+    }
+    
+    private static inline function get_identity():DualQuaternion
+    {
+        return new DualQuaternion(Quaternion.identity, Quaternion.zero);
+    }
+    
+    private inline function get_normal():DualQuaternion
+    {
+        var mag = dot(this, this);
+        if( mag < 0 ) return null;
+        var oneOver = 1/mag;
+        return new DualQuaternion(this.real*oneOver, this.dual*oneOver);
+    }
+    
+    /**
+     * Normalize the quaternion in-place.
+     * 
+     * @return  The modified object.
+     */
+    public inline function normalize(): DualQuaternion
+    {
+        var mag = dot( this, this );
+        if( mag < 0 ) return null;
```

Comment:
Duplicates code with get_normal. Suggestion: call .clone().normalize() in get_normal. I actually don't do this correctly in my Quaternion code, so normal does fail on 0-length quaternions (oops).